File: dotnet/TellDontAskKata/TellDontAskKata/UseCase/OrderCreationUseCase.cs
using System;
using System.Collections.Generic;
using TellDontAskKata.Domain;
using TellDontAskKata.Repository;

namespace TellDontAskKata.UseCase
{
    public class OrderCreationUseCase
    {
        private readonly IOrderRepository orderRepository;
        private readonly IProductCatalog productCatalog;

        public OrderCreationUseCase(IOrderRepository orderRepository,
            IProductCatalog productCatalog)
        {
            this.orderRepository = orderRepository;
            this.productCatalog = productCatalog;
        }

        public void Run(SellItemsRequest request)
        {
            var order = new Order
            {
                Status = OrderStatus.Created,
                Items = new List<OrderItem>(),
                Currency = "EUR",
                Total = 0.0m,
                Tax = 0.0m
            };

            foreach (var itemRequest in request.Requests)
            {
                var product = productCatalog.GetByName(itemRequest.ProductName);
                if (product == null)
                {
                    throw new UnknownProductException();
                }
                else

                {

                    var unitaryTax = Math.Round(product.Price / 100 * product.Category.TaxPercentage, 2,
                        MidpointRounding.AwayFromZero);

                    var unitaryTaxAmount = Math.Round(product.Price + unitaryTax, 2, MidpointRounding.AwayFromZero);

                    var taxedAmount = Math.Round(unitaryTaxAmount * itemRequest.Quantity, 2,
                        MidpointRounding.AwayFromZero);

                    var taxAmount = Math.Round(unitaryTax * itemRequest.Quantity, 2, MidpointRounding.AwayFromZero);

                    var orderItem = new OrderItem
                    {
                        Product = product,
                        Quantity = itemRequest.Quantity,
                        Tax = taxAmount,
                        TaxedAmount = taxedAmount
                    };

                    order.Items.Add(orderItem);
                    order.Total += taxedAmount;
                    order.Tax += taxAmount;
                }
            }
            orderRepository.Save(order);
        }
    }
}

File: scala/src/main/scala/ordershipping/usecase/OrderCreationUseCase.scala
package ordershipping.usecase

import ordershipping.domain.{Order, OrderItem, OrderStatus}
import ordershipping.repository.{OrderRepository, ProductCatalog}

import scala.collection.mutable

class OrderCreationUseCase(
    val orderRepository: OrderRepository,
    val productCatalog: ProductCatalog
) {
  def run(request: SellItemsRequest): Unit = {
    val order = new Order(
      total = 0d,
      currency = "EUR",
      items = mutable.MutableList.empty,
      tax = 0d,
      status = OrderStatus.Created,
      id = 1
    )

    for (itemRequest <- request.requests) {
      val product = productCatalog.getByName(itemRequest.productName)

      if (product.isEmpty)
        throw new UnknownProductException
      else {
        product.foreach(p => {
          val unitaryTax =
            roundAt(2)((p.price / 100) * p.category.taxPercentage)
          val unitaryTaxedAmount = roundAt(2)(p.price + unitaryTax)
          val taxedAmount =
            roundAt(2)(unitaryTaxedAmount * itemRequest.quantity)
          val taxAmount = roundAt(2)(unitaryTax * itemRequest.quantity)

          val orderItem = new OrderItem(
            product = p,
            quantity = itemRequest.quantity,
            taxedAmount = taxedAmount,
            tax = taxAmount
          )

          order.items += orderItem
          order.total += taxedAmount
          order.tax += taxAmount
        })
      }
    }
    orderRepository.save(order)
  }

  private def roundAt(p: Int)(n: Double): Double = {
    val s = math pow (10, p)
    (math round n * s) / s
  }
}
